automation/taskcluster/decision_task.py

```diff
@@ -214,36 +233,106 @@ def _get_release_gradle_tasks(module_name, is_snapshot):
 def release(artifacts_info, version, is_snapshot, is_staging):
     version = components_version() if version is None else version
 
+    docker_images_tasks, build_docker_image_task_id = create_docker_tasks_if_needed(is_release=True)
     build_tasks = {}
     wait_on_builds_tasks = {}
     beetmover_tasks = {}
     other_tasks = {}
-    wait_on_builds_task_id = taskcluster.slugId()
+    wait_on_builds_task_id = craft_new_task_id()
 
     for artifact_info in artifacts_info:
-        build_task_id = taskcluster.slugId()
+        build_task_id = craft_new_task_id()
         module_name = _get_gradle_module_name(artifact_info)
         build_tasks[build_task_id] = BUILDER.craft_build_task(
             module_name=module_name,
             gradle_tasks=_get_release_gradle_tasks(module_name, is_snapshot),
             subtitle='({}{})'.format(version, '-SNAPSHOT' if is_snapshot else ''),
             run_coverage=False,
             is_snapshot=is_snapshot,
-            artifact_info=artifact_info
+            artifact_info=artifact_info,
+            build_docker_image_task_id=build_docker_image_task_id,
         )
 
-        beetmover_tasks[taskcluster.slugId()] = BUILDER.craft_beetmover_task(
+        beetmover_tasks[craft_new_task_id()] = BUILDER.craft_beetmover_task(
             build_task_id, wait_on_builds_task_id, version, artifact_info['artifact'],
             artifact_info['name'], is_snapshot, is_staging
         )
 
-    wait_on_builds_tasks[wait_on_builds_task_id] = BUILDER.craft_wait_on_builds_task(build_tasks.keys())
+    wait_on_builds_tasks[wait_on_builds_task_id] = BUILDER.craft_wait_on_builds_task(
+        build_tasks.keys(), build_docker_image_task_id
+    )
 
     if is_snapshot:     # XXX These jobs perma-fail on release
         for craft_function in (BUILDER.craft_detekt_task, BUILDER.craft_ktlint_task, BUILDER.craft_compare_locales_task):
-            other_tasks[taskcluster.slugId()] = craft_function()
+            other_tasks[craft_new_task_id()] = craft_function(build_docker_image_task_id)
+
+    return (docker_images_tasks, build_tasks, wait_on_builds_tasks, beetmover_tasks, other_tasks)
+
+
+def create_docker_tasks_if_needed(is_release=False):
+    tasks = {}
+    docker_folders = [
+        os.path.join(DOCKER_FOLDER, f)
+        for f in os.listdir(DOCKER_FOLDER)
+        if os.path.isdir(os.path.join(DOCKER_FOLDER, f))
+    ]
+
+    for folder in docker_folders:
+        folder_name = os.path.basename(folder)
+        folder_hash = dirhash(folder, 'sha256')
+        try:
+            task_id = get_existing_task_in_index(TRUST_LEVEL, folder_name, folder_hash)
```

Comment:
Minor nit: readability might be improved if `get_existing_task_in_index` returned `None` (rather than throwing an exception) if an existing task couldn't be found.

This is for two reasons:
1. Otherwise, a potentially-unrelated `ValueError` might be handled as a "task couldn't be found"
2. It disrupts code flow a little bit: to track "normal, expected" behaviour that happens after this function is called, the reader needs to follow the subsequent code _and_ jump to the `except`.
    * By handling the `None` here, it's clear how the branch happens here
    * Additionally, by leaving real failures/unexpected situations to exceptions, we can establish a convention that `except` sections are for handling edge cases that "shouldn't happen in real operation"